agent/health_endpoint.go

```diff
@@ -219,14 +219,19 @@ func (s *HTTPHandlers) healthServiceNodes(resp http.ResponseWriter, req *http.Re
 		return nil, nil
 	}
 
-	useStreaming := s.agent.config.UseStreamingBackend && args.MinQueryIndex > 0
+	useStreaming := s.agent.config.UseStreamingBackend && args.MinQueryIndex > 0 && !args.Ingress
 	args.QueryOptions.UseCache = s.agent.config.HTTPUseCache && (args.QueryOptions.UseCache || useStreaming)
 
 	if args.QueryOptions.UseCache && useStreaming && args.Source.Node != "" {
 		return nil, BadRequestError{Reason: "'near' query param can not be used with streaming"}
 	}
 
-	out, md, err := s.agent.rpcClientHealth.ServiceNodes(req.Context(), args)
+	healthClient := s.agent.rpcClientHealth
+	if args.Ingress {
+		healthClient = s.agent.rpcClientHealthNoStreaming
+	}
```

Comment:
The purpose of the `rpcclient` is to abstract this logic from the different places that use it.  The DNS server also makes these queries and would not be fixed by this change.

I think what needs to happen is restoring the `CacheNameConnect` (or maybe `CacheNameNoConnect` instead) field the to rpc client. It was removed in https://github.com/hashicorp/consul/pull/9152/commits/1a764553c028be1078373fb51e71c71c4e87802a because we no longer needed it, but I guess we still do for this case.

Reply:
:+1:

Reply:
Updated